### PR TITLE
refactor(trie): abstract database implementation with an interface

### PR DESCRIPTION
**Pull request by faustbrian**
Resubmission of https://github.com/ethereumjs/ethereumjs-monorepo/pull/1895 with squashed history and rebase on `develop`

### CODE REVIEW

**Comment by holgerd77** [comment]
Rebased this via UI after testing locally.

**Review by holgerd77 — APPROVED** [review]
LGTM, great great work, thanks so much! 😄 

Will merge.

**Comment by holgerd77** [comment]
Could you outline a potential API usage (instantiation) in 4-5 lines?

**Comment by holgerd77** [comment]
Ok, thanks, will ask within the team! 🙂

**Comment by holgerd77** [comment]
(I guess chances are 50:50)

File: packages/trie/benchmarks/index.ts [hunk 1]
```diff
@@ -1,45 +1,50 @@
...
-  .add('1k-9-32-ran', async () => {
-    await runTrie(9, false)
-  })
-  .add('1k-1k-32-ran', async () => {
-    await runTrie(1000, false)
-  })
-  .add('1k-1k-32-mir', async () => {
-    await runTrie(1000, true)
-  })
+for (const [name, DB] of Object.entries({ MemoryDB, LevelDB })) {
```
Comment (ryanio) [inline packages/trie/benchmarks/index.ts:8]: hey thanks for the great PR, continued work and interest!

I'm not sure if it's worth doubling the number of benchmarks run here on every PR. What we could do instead is a one time profile difference between the two and record the results in the readme, and save the script so we could add in the future like lmdb comparison.

**Comment by holgerd77** [comment]
> @holgerd77 created a draft of the changes to show what it would actually entail - [faustbrian@da41ce5](https://github.com/faustbrian/ethereumjs-monorepo/commit/da41ce53a0b1b5b124751eec4aa17c5417706d6a)

Thanks a lot for all this pro-active engagement, that's really great!! 👍 

I guess we do not have an agreement yet, opinions are somewhat mixed and reserved, will keep you updated!

File: packages/trie/benchmarks/index.ts (hunk 1 above)
Comment (ryanio) [reply packages/trie/benchmarks/index.ts:8]: Sure a BENCHMARK.md sounds good too, you can link to it from the readme so people can discover it. There is already a trie-benchmarks GitHub action so my intention was to not need to run the comparison every time.

**Comment by holgerd77** [comment]
> @holgerd77 created a draft of the changes to show what it would actually entail - [faustbrian@da41ce5](https://github.com/faustbrian/ethereumjs-monorepo/commit/da41ce53a0b1b5b124751eec4aa17c5417706d6a)

Hi there, to follow-up: still no clear opinion crystalizing out. Do you see (have yourself) some substantial practical gains from this separation or is this rather some "we should do it since this is how things should be?" thing?

(To be clear: I (and I guess I can say: we) would also very much value the second argument, however in the current situation first argument would give a stronger reasoning to do since we do have to limit ourselves right now and from the "it would be structurally cleaner to do x" perspective we still have a lot on the plate which we would/could/would want to realize at some point (but also can't/won't do right now for very much practical/procedural reasons))

**Comment by holgerd77** [comment]
(and so be totally transparent on this: my current tendency is to rather *not* do since this does feel as overloading to me, this this current release round is just already demanding *a lot* from the team due to it's dramatically large scope it already has)

**Comment by holgerd77** [comment]
> Pretty much "how it should be done" but also practical gains resulting from that because then the trie package can be used in browsers which would resolve #1056. Alternatively you could also upgrade `level/level` to `^8.0` and that way get an easy way to add browser support without removing LevelDB.

Thanks, we have various `level` usages in our package suite - 4-6 or so and all on a similar version range - e.g. also as a block data store in the blockchain package, in the client.

Can you judge on the effort to then bring *all* of this to level 8? This might then also be a way here, would also fit well in the scope of our release process, to do such an upgrade right now.

**Comment by holgerd77** [comment]
> I upgraded it in of my own applications and that only took an hour with testing if everything worked but I can work on a PR tomorrow and submit a draft during the weekend so you can see what it looks like.

Uh. If it is not too much of a burden too you this would be great! 💯

Thanks a lot for all these improvement offers - also for being flexible what to integrate. So when you do a `"level` search with the `packages` scope, you get 8 occurrences of `level` and `level`-related (`level-mem`) packages.

Is this strictly a breaking change, or would old DBs still work, do you know that (we would nevertheless do towards `develop`, just asking)?

Also, in case you do you might want to wait for #1928, this is our very last rebase of `develop` towards the `master` branch, from that on we will have no more `master` changes and `develop` will get merged into `master` soon (days, sometime next week).

**Comment by holgerd77** [comment]
@faustbrian Hi there, we have now merged in `develop` into `master`. So if you find some time you can submit a PR directly towards `master` now 🙂.

According to our timeline from #1914 a feature freeze for the v6 `master` is intended to be applied on June 12. We are - mostly - in time, but this might also get 1-2 more days though.

**Pull request merged**